src/app/js/lib/components/ButtonWithStatus.js

```diff
@@ -1,47 +1,93 @@
 import React from 'react';
+import classnames from 'classnames';
 import PropTypes from 'prop-types';
 import { lightGreen, red } from '@material-ui/core/colors';
-import { CircularProgress, Button } from '@material-ui/core';
+import { CircularProgress, LinearProgress, Button } from '@material-ui/core';
 import Warning from '@material-ui/icons/Warning';
 import Success from '@material-ui/icons/Done';
-
-const getIcon = (error, loading, success) => {
-    if (loading) return <CircularProgress variant="indeterminate" size={20} />;
-    if (error) return <Warning color={red[400]} />;
-    if (success) return <Success color={lightGreen.A400} />;
-    return null;
-};
+import { makeStyles } from '@material-ui/styles';
+import theme from '../../theme';
 
 const ButtonWithStatus = ({
     raised,
     error,
     loading,
     disabled,
     success,
+    progress,
+    target,
+    className,
     ...props
-}) =>
-    raised ? (
-        <Button
-            variant="contained"
-            disabled={disabled || loading}
-            startIcon={getIcon(error, loading, success)}
-            {...props}
-        />
-    ) : (
-        <Button
-            variant="text"
-            disabled={disabled || loading}
-            startIcon={getIcon(error, loading, success)}
-            {...props}
-        />
+}) => {
+    const classes = useStyles();
+    return (
+        <div className={classes.container}>
+            {raised ? (
+                <Button
+                    variant="contained"
+                    className={classnames(className, {
+                        [classes.loadingProgress]: loading && target,
+                    })}
+                    disabled={disabled || loading}
+                    startIcon={getIcon(error, success, loading, target)}
+                    {...props}
+                />
+            ) : (
+                <Button
+                    variant="text"
+                    disabled={disabled || loading}
+                    startIcon={getIcon(error, success, loading, target)}
+                    {...props}
+                />
+            )}
+            {loading && target ? (
+                <LinearProgress
+                    // className={classes.progress}
+                    classes={{
+                        root: classes.progress,
+                        colorPrimary: classes.colorPrimary,
+                        barColorPrimary: classes.barColorPrimary,
+                    }}
+                    variant="determinate"
+                    value={target ? (progress / target) * 100 : 0}
+                />
+            ) : (
+                undefined
+            )}
+        </div>
     );
+};
+const useStyles = makeStyles({
```

Comment:
More a global question here, do we have conventions for the order of elements in a component? 
Because sometimes the style comes first, sometimes after the render etc...

Reply:
We should, what do you prefer, I'm not a very consistent guy... So I'll stick to what you prefer ;)